DreamFactory sample update for Swift 3.0

eelfner opened the pull request:
I updated the Swift sample code for Swift 3.0. Basic operation verified. Did not do complete regression tests. I was updating some other projects and thought I'd go ahead and do a first pass on this for you. Let me know if you have any questions.

df-admin commented:
Thanks Eric! We'll test it out and merge if everything looks ok.

shingsoso commented:
Is the project stop updating or what lol ?

eelfner commented:
@shingsoso I have not worked with DF for quite a while. I had fixed compile errors in this project long ago as a very small paid project. When Swift 3 came out, I went ahead and updated and created this pull request just to be helpful which looks like was never merged. Hum?

A more interesting project is my more complete rewrite of this iOS sample [here](https://github.com/eelfner/DreamFactory-EZ-iOS), but it also needs refreshing. I think DF is a great concept, but unsure how widely it has been accepted.

shingsoso commented:
@eelfner  Ooops , thanks for the linked sample rewrote by you ! Can I use it as reference for my app developement ? Appreciate what you did !!!

eelfner commented:
Sure you can use it. Hope it is helpful. I'm sure it needs some updates, and please note that my DF server is surely gone by now so you'll need to setup your own. I went ahead and added MIT license which basically allows you to do anything you want.

zerox1212 commented:
This should really be taken and also moved up to Swift 4. I had to do this on my own app that was based on this library and it would save a lot of people frustration (especially non-iOS programmers). There are few things int he JSON serializing/deserializing that are tricky to fix.

eelfner commented:
@zerox1212 it is hard for me to know what is going on at Dream Factory. I love the idea of what they are doing and did some paid (very little) and more some free work for them a while back. I don't think I'm going to continue to maintain this example since they have not taken my last update and I need to focus on more lucrative projects.

I do have another Dream Factory git project that is a re-work of this demo. My goal was to simplify the coding and UI flow without a total overhaul. I just updated it for Swift 4. Also, updated it with new Dream Factory demo server constants, so it should run out-of-the-box until that server app gets reset.

You can find it: https://github.com/eelfner/DreamFactory-EZ-iOS

todda48 commented:
@eelfner I really appreciate the help you have given us on this. I know some of the sample apps need updating, and we are trying to get to them as time permits. Sorry for the delay.

zerox1212 commented:
@eelfner I updated my code, but I'm not very good with Swift yet.

Can you give me any clues about how to get the CustomNavBar to work in swift 4? After I updated the buttons in the nav bar don't fire events. It seems to be something with using `#selector`. My selector functions are marked with `@objc` but they still don't fire. They don't hit break points and there are no errors.

Any idea?

eelfner commented:
@zerox1212 I went ahead and debugged this. This code is very non-standard iOS code and will be problematic. I do not recommend using it as a model. That said, something change in the SDK and this issue has nothing to do with Swift 4 and is possibly an internal issue. Anyway, it seems a simple 1 line addition will fix this. Add the line tb.layoutIfNeeded() in AppDelegate.swift after the CustomNavBar is created.
  
```
func application(_ application: UIApplication, didFinishLaunchingWithOptions...
        
        // Build persistent navigation bar
        let tb = CustomNavBar()
        tb.layoutIfNeeded()     // Required after latest SDK
```

This codebase creates this custom nav bar, but the add buttons get buried under the content view until the layout is done.

zerox1212 commented:
@eelfner Thank you so much! I'm an Android programmer and I suck at working in iOS! If your solution works it will be great!!!

For my own learnings, how did you troubleshoot this?

eelfner commented:
@zerox1212 This was a pretty tricky issue and all the extra non-standard code in this project combined to make it quite a challenge. 

- Having done some other Swift 4 conversions, I looked at the `@objc` things which you already did and are a common issue. 
- The break came when I realized the button was not giving any visual indication of being clicked. 
- This led me to an investigation of the visual hierarchy and I saw the button was under the content view of the toolbar. Odd! 
- I found if I disabled user interaction on that overlaid content view, the buttons started working again.
- If figured it with some issue with the replacement of the standard navigation controller toolbar with the custom one. I coded around the setup to get it too work, but it was a bit involved so I looked for a better solution.
- Then found a StackOverflow posting of a similar issue with iOS11 SDK, that led me to the easy answer. 
- Tore out the other code, and the one line fix seemed to work.

YMMV... Could be other issues... Best, Eric

zerox1212 commented:
If you are on Codementor please contact me andrewj@zeroentity.com. Thanks again.
